polkadot/xcm/xcm-builder/src/asset_exchange.rs

```diff
@@ -0,0 +1,146 @@
+		> + QuoteExchangePrice<Balance = u128, AssetKind = Fungibles::AssetId>,
+	Fungibles: fungibles::Balanced<AccountId, Balance = u128>,
+	Matcher: MatchesFungibles<Fungibles::AssetId, Fungibles::Balance>,
+{
+	fn exchange_asset(
+		_: Option<&Location>,
+		give: AssetsInHolding,
+		want: &Assets,
+		_: bool,
+	) -> Result<AssetsInHolding, AssetsInHolding> {
+		let give_asset = give.fungible_assets_iter().next().ok_or_else(|| {
+			log::trace!(
+				target: "xcm::FungiblesPoolAdapter::exchange_asset",
+				"No fungible asset was in `give`.",
+			);
+			give.clone()
+		})?;
+		let want_asset = want.get(0).ok_or_else(|| {
+			log::trace!(
+				target: "xcm::FungiblesPoolAdapter::exchange_asset",
+				"No asset was in `want`.",
+			);
+			give.clone()
+		})?;
+		let (give_asset_id, balance) =
+			Matcher::matches_fungibles(&give_asset).map_err(|error| {
+				log::trace!(
+					target: "xcm::FungiblesPoolAdapter::exchange_asset",
+					"Could not map XCM asset {:?} to FRAME asset. Error: {:?}",
+					give,
+					error,
+				);
+				give.clone()
+			})?;
+		let (want_asset_id, want_amount) =
+			Matcher::matches_fungibles(&want_asset).map_err(|error| {
+				log::trace!(
+					target: "xcm::FungiblesPoolAdapter",
+					"Could not map XCM asset {:?} to FRAME asset. Error: {:?}",
+					want,
+					error,
+				);
+				give.clone()
+			})?;
+
+		// We have to do this to convert the XCM assets into credit the pool can use.
+		let swap_asset = give_asset_id.clone().into();
+		let credit_in = Fungibles::issue(give_asset_id, balance);
+		log::trace!(target: "xcm", "Credit in: {:?}", credit_in.peek());
+
+		// Do the swap.
+		let (credit_out, credit_change) =
+			<AssetConversion as SwapCredit<_>>::swap_tokens_for_exact_tokens(
+				vec![swap_asset, want_asset_id],
+				credit_in,
+				want_amount,
+			)
+			.map_err(|(credit_in, _)| {
+				drop(credit_in);
+				give.clone()
+			})?;
+
+		log::trace!(target: "xcm", "Credit out: {:?}, Credit change: {:?}", credit_out.peek(), credit_change.peek());
+		debug_assert!(credit_change.peek() == Zero::zero());
+
+		let resulting_asset: Asset = (want_asset.id.clone(), credit_out.peek()).into();
+		Ok(resulting_asset.into())
+	}
+
+	fn quote_exchange_price(asset1: &Asset, asset2: &Asset, _: bool) -> Option<u128> {
+		let (asset1_id, _) = Matcher::matches_fungibles(asset1)
+			.map_err(|error| {
+				log::trace!(
+					target: "xcm::FungiblesPoolAdapter::quote_exchange_price",
+					"Could not map XCM asset {:?} to FRAME asset. Error: {:?}.",
+					asset1,
+					error,
+				);
+				()
+			})
+			.ok()?;
+		let (asset2_id, desired_asset2_amount) = Matcher::matches_fungibles(asset2)
+			.map_err(|error| {
+				log::trace!(
+					target: "xcm::FungiblesPoolAdapter::quote_exchange_price",
+					"Could not map XCM asset {:?} to FRAME asset. Error: {:?}.",
+					asset2,
+					error,
+				);
+				()
+			})
+			.ok()?;
+		let necessary_asset1_amount =
+			<AssetConversion as QuoteExchangePrice>::quote_price_tokens_for_exact_tokens(
```

Comment:
returns option, no need for tmp var

Reply:
Now I use temp var because I decide on the function to call based on `maximal`